### ISSUE TITLE
Large JSONRPC 'id' field gets converted to negative integer

### ISSUE THREAD

**Issue by fumecow** [rsksmart/rskj#169]
I've encountered a bug with the low-level RPC communication to / from our private RSKj node caused by the way Truffle sometimes uses timestamps for the 'id' field in JSONRPC requests.

The RSKj node appears to convert the large 'id' field to a signed integer internally so the RPC response has a negative 'id' field and the result is our test fails with no events reported.

The packet capture is below.  Note how the id for normal requests is in the 240 range, then we get 1501691352102 for the eth_getLogs call, and the response has id = -1547201498

Wed Aug 02 2017 10:29:12 GMT-0600 (MDT) {"jsonrpc":"2.0","method":"eth_getTransactionReceipt","params":["0x196f411b62566b5ec5f34c0baff34035d650cb488b9c8fa5a725a5c00276812d"],"id":240}
Wed Aug 02 2017 10:29:12 GMT-0600 (MDT) {"jsonrpc":"2.0","id":240,"result":{"transactionHash":"0x196f411b62566b5ec5f34c0baff34035d650cb488b9c8fa5a725a5c00276812d","transactionIndex":"0x0","blockHash":"0xf0dc891cd7d1d6dada639e75aa6c2c91a4e247d03aa8408b41ece370e4ca8925","blockNumber":"0x922","cumulativeGasUsed":"0x8f6a","gasUsed":"0x8f6a","contractAddress":null,"logs":[],"from":"0xcd2a3d9f938e13cd947ec05abc7fe734df8dd826","to":"0x3056d0b2379c86e6be6771c34f39f5affb65d855","root":"0x8c188565840989875932e0125377dd4ff6ddea7c7c09ac1c0f45e2cd78868935"}}
Wed Aug 02 2017 10:29:12 GMT-0600 (MDT) {"jsonrpc":"2.0","method":"eth_getBalance","params":["0x0179d39e7c0bc25bf495f706c893c78d39b3421f","latest"],"id":27}
Wed Aug 02 2017 10:29:12 GMT-0600 (MDT) {"jsonrpc":"2.0","id":27,"result":"0x1c6bc7fae274ae8f"}
Wed Aug 02 2017 10:29:12 GMT-0600 (MDT) {"jsonrpc":"2.0","method":"eth_getLogs",**"id":1501691352102**,"params":[{"fromBlock":"0x91b"}]}
Wed Aug 02 2017 10:29:12 GMT-0600 (MDT) {"jsonrpc":"2.0",**"id":-1547201498**,"result":[]}
Wed Aug 02 2017 10:29:12 GMT-0600 (MDT) {"jsonrpc":"2.0","method":"evm_revert","id":1501691352105,"params":["0x3"]}
Wed Aug 02 2017 10:29:12 GMT-0600 (MDT) {"jsonrpc":"2.0","id":-1547201495,"result":true}
Wed Aug 02 2017 10:29:12 GMT-0600 (MDT) {"jsonrpc":"2.0","method":"evm_snapshot","id":1501691352108}
Wed Aug 02 2017 10:29:12 GMT-0600 (MDT) {"jsonrpc":"2.0","id":-1547201492,"result":"0x4"}
Wed Aug 02 2017 10:29:12 GMT-0600 (MDT) {"jsonrpc":"2.0","method":"eth_accounts","params":[],"id":28}

**Comment by tinchou** [comment]
Initial investigation points us to a possible bug in jsonrpc4j. Objects of type LongNode return true to the isIntegralNumber check in the following line, and then get casted to int: https://github.com/briandilley/jsonrpc4j/blob/master/src/main/java/com/googlecode/jsonrpc4j/JsonRpcBasicServer.java#L858

Next steps:

1. Reproduce the bug in upstream code
1. Look for a workaround while we wait for upstream to release a new version with the fix

**Comment by tinchou** [comment]
To reproduce the bug:

```bash
curl -X POST\
     -H "Content-Type: application/json"\
     -d '{"jsonrpc":"2.0","method":"eth_getLogs","id":1501691352102,"params":[{"fromBlock":"0x91b"}]}'\
     http://localhost:4444
```

Actual response:

```bash
{"jsonrpc":"2.0","id":-1547201498,"result":[]}
```

Expected response:

```bash
{"jsonrpc":"2.0","id":1501691352102,"result":[]}
```

This is fixed with `jsonrpc4j:1.5.1`. We'll go through security review to make this update and keep you posted.

@fumecow if you're compiling the project yourself, you can apply this change and test the fix: https://github.com/rsksmart/rskj/commit/119dfb6821344a4b09115b22ad5517c782e607f7

**Comment by fumecow** [comment]
thanks for the update.  will take a look...

**Comment by aeidelman** [comment]
Fixed in bamboo v0.3.0
